File: 84_Largest_Rectangle_in_Histogram/Solution.java
import java.util.*;

public class Solution {
    public int largestRectangleArea(int[] height) {
        int len = height.length;
        Stack<Integer> s = new Stack<Integer>();
        int maxArea = 0;
        for(int i = 0; i < len; i++){
            // int h = (i == len ? 0 : height[i]);
            int h = height[i];
            if(s.isEmpty() || h >= height[s.peek()]){
                s.push(i);
            }else{
                int tp = s.pop();
                maxArea = Math.max(maxArea, height[tp] * (s.isEmpty() ? i : i - 1 - s.peek()));
                i--;
            }
        }
        return maxArea;
    }

    public static void main(String[] args){
        Solution s = new Solution();
        // System.out.println(s.largestRectangleArea(new int[]{2,1,5,6,2,3}));
        System.out.println(s.largestRectangleArea(new int[]{2,1,2}));
    }
}

File: 84_Largest_Rectangle_in_Histogram/Solution.py
class Solution:
    def largestRectangleArea(self, heights):
        if heights == None or len(heights) == 0:
            return 0


        n = len(heights)        
        stck = []
        mx = 0
        for i in range(n):
            j = i
            t = 0
            while j>-1 and heights[i] <= heights[j]:
                heights[j] = heights[i]
                t += 1
                j -= 1

            mx = max(mx, t* heights[i])
        for i in range(n):
            mx = max(mx, heights[i]*(n-i))

        return mx
        # for i in range(n):
        #     t = 0
        #     while len(stck) !=0 and heights[i] <= stck[-1]:
        #         t += 1
        #         stck.pop()
        #     if len(stck) > 0:
        #         mx = max(mx, (t+1)* heights[i], (t+2)*stck[-1])
        #     else:
        #         mx = max(mx, (t+1)*heights[i])
        #     stck.append(heights[i])
        # return mx

s = Solution()
print(s.largestRectangleArea([2,1,5,6,2,3]))
